src/pykeen/nn/emb.py

```diff
@@ -580,3 +582,313 @@ def forward(
+            be a sequence / list.
+        :param dims:
+            The hidden dimensions to use. If None, defaults to the embedding dimension of the base representations.
+            If an integer, is the same for all layers. The last dimension is equal to the output dimension.
+        :param layer_kwargs:
+            Additional key-word based parameters passed to the individual layers; cf. CompGCNLayer.
+        """
+        super().__init__()
+        # TODO: Check
+        assert triples_factory.create_inverse_triples
+        self.entity_representations = embedding_specification.make(
+            num_embeddings=triples_factory.num_entities,
+        )
+        self.relation_representations = embedding_specification.make(
+            num_embeddings=2 * triples_factory.real_num_relations,
+        )
+        input_dim = self.entity_representations.embedding_dim
+        assert self.relation_representations.embedding_dim == input_dim
+
+        # hidden dimension normalization
+        if dims is None:
+            dims = input_dim
+        if isinstance(dims, int):
+            if num_layers is None:
+                raise ValueError
+            else:
+                dims = [dims] * num_layers
+        if len(dims) != num_layers:
+            raise ValueError(
+                f"The number of provided dimensions ({len(dims)}) must equal the number of layers ({num_layers}).",
+            )
+        self.output_dim = dims[-1]
+
+        # Create message passing layers
+        layers = []
+        for input_dim, output_dim in zip(itertools.chain([input_dim], dims), dims):
+            layers.append(CompGCNLayer(
+                input_dim=input_dim,
+                output_dim=output_dim,
+                **(layer_kwargs or {}),
+            ))
+        self.layers = nn.ModuleList(layers)
+
+        # register buffers for adjacency matrix; we use the same format as PyTorch Geometric
+        # TODO: This always uses all training triples for message passing
+        self.register_buffer(name="edge_index", tensor=triples_factory.mapped_triples[:, [0, 2]].t())
+        self.register_buffer(name="edge_type", tensor=triples_factory.mapped_triples[:, 1])
+
+        # initialize buffer of enriched representations
+        self.enriched_representations = None
+
+    def post_parameter_update(self) -> None:  # noqa: D102
+        # invalidate enriched embeddings
+        self.enriched_representations = None
+
+    def forward(
+        self,
+    ) -> Tuple[Embedding, Embedding]:
+        """Compute enriched representations."""
+        if self.enriched_representations is None:
+            x_e = self.entity_representations()
+            x_r = self.relation_representations()
+            # enrich
+            for layer in self.layers:
+                x_e, x_r = layer(x_e=x_e, x_r=x_r, edge_index=self.edge_index, edge_type=self.edge_type)
```

Comment:
where do self.edge_index and self.edge_type come from?

Reply:
From the `register_buffer` in `__init__`.

```python
self.register_buffer(name="edge_index", tensor=tensor)
```

roughly is equivalent to 
```python
self.edge_index = tensor
```
with the difference that torch knows that this attribute is a tensor which should be moved along the module (i.e. `self.to("cuda")` will move all registered parameters and buffers to GPU).

Reply:
cf. https://pytorch.org/docs/stable/generated/torch.nn.Module.html#torch.nn.Module.register_buffer

Reply:
ah yes, my primitive concepts of python programming have been steamrolled by torch magic

Reply:
With parameters, we have a nicer solution, since

```python
self.w = nn.Parameter(...)
```

automatically registers the parameter without having to call [register_parameter](https://pytorch.org/docs/stable/generated/torch.nn.Module.html#torch.nn.Module.register_parameter)